Add ability to pass objects from test_description to tests

pbaughman opened the pull request:
### Description
To facilitate writing test factories (and to clean up the module scope of tests a little bit), we need a way to pass information from the `generate_test_description` function over to the tests.  Previously you could do this by [declaring things at the module level](https://github.com/ApexAI/apex_rostest/blob/master/apex_launchtest/examples/good_proc.test.py#L36-L39) but that's a little gross, and it doesn't make it very easy to write a factory class that makes test descriptions and automatically makes the corresponding tests that go along with.

To make this cleaner, the generate_test_description function can optionally return a second object - a dictionary.  ~~Each item in the dictionary will be injected into the individual tests as though the object was an attribute with the same name as the dictionary key.~~  Each item in the dictionary will be passed to the individual test cases if they're declared with an argument with the same name.

This will let us write factory methods that return a complicated launch descriptions and pass parts of the complicated launch description transparently to factory tests.

#### Notes for Reviewers
I left good_proc.test.py the way it was because the examples are also test cases, and I want to make sure we don't regress the case where a test description only returns one object

example_test_context.test.py shows the new functionality, and has some simple tests that use the launch context

#### Todo:
 - ~~_bind_test_context_to_tests should make sure the attribute name doesn't already exist so we're not overwriting important test functionality.  This should probably be an error~~
 - [x] Merge https://github.com/ApexAI/apex_rostest/pull/22 and rebase this branch, then change the PR to target master

Commit:
Add ability to pass objects from test_description to tests

pbaughman commented:
@hidmic This _kind of_ does what you were talking about in https://github.com/ApexAI/apex_rostest/issues/20 but doesn't get you the launch context.  It will reduce the amount of stuff that lives at the module level scope, though.

pbaughman commented on apex_launchtest/examples/example_test_context.test.py:
```diff
@@ -0,0 +1,80 @@
+
+    # Items in this dictionary will be added to the test cases as an attribute based on
+    # dictionary key
+    test_context = {
```
Here's the new feature in action.  The 2nd value returned by generate_test_description can be a dictionary that will be injected into the tests.  In this case 'self.dut' and 'self.int_val'

hidmic replied:
Looks promising! :)

hidmic commented on apex_launchtest/apex_launchtest/apex_runner.py:
```diff
@@ -177,6 +196,10 @@ def _print_process_output_summary(self):
 
+    def _bind_test_context_to_tests(self, context, test_suite):
+        for k, v in context.items():
+            self._give_attribute_to_tests(v, k, test_suite)
```
@pbaughman hmm, why not passing them as arguments for the tests instead? wouldn't that preclude attribute clashes?

pbaughman replied:
@hidmic I don't have a ton of control over what's passed into the tests as arguments, although maybe someone with a better understanding of unittest can point me in the right direction.  We delegate that to the unittest.TextTestRunner [here](https://github.com/ApexAI/apex_rostest/blob/3233d7992c48423af25781270650935b8efaf1f3/apex_launchtest/apex_launchtest/apex_runner.py#L161) and [here](https://github.com/ApexAI/apex_rostest/blob/3233d7992c48423af25781270650935b8efaf1f3/apex_launchtest/apex_launchtest/apex_runner.py#L129)

The [runner calls run](https://github.com/python/cpython/blob/dc078947a5033a048d804e244e847b5844734439/Lib/unittest/runner.py#L176) on the test suite which [calls run](https://github.com/python/cpython/blob/dc078947a5033a048d804e244e847b5844734439/Lib/unittest/suite.py#L122) on the test case which [executes the actual test method](https://github.com/python/cpython/blob/master/Lib/unittest/case.py#L642) and doesn't pass arguments

pbaughman replied:
Thinking about this on the bus last night. . .

Maybe I could pre-process the list of test methods and bind the arguments before I pass them into TextTestRunner.run

I think there's a functools method for that

Commit:
Pass test context as arguments to test cases

  - If we pass the test_context dictionary items as kwargs to the tests, we don't
    need to worry about them colliding with other fields on the unittest.TestCase.
    Someone could still name a TestContext item "self" but I feel like they deserve
    what they get if they do that.
  - Refactor the method that iterates all test cases so it can be used to give
    attributes to test, as well as the new bind_test_context_to_tests method

Commit:
Fix doc comments

pbaughman commented:
@hidmic Ok, what about this?  We can partially bind arguments to the test methods before we pass them to the runner.  I had to touch some underbar variables in unittest.TestCase to do it, but YOLO.  Bad stuff will happen if someone adds something to the test context named 'self' but like my doctor says "if it hurts when you do that, don't do that"

pbaughman commented on apex_launchtest/apex_launchtest/apex_runner.py:
```diff
@@ -177,20 +197,40 @@ def _print_process_output_summary(self):
+
+            # Replace the test with a functools.partial that has the arguments
+            # provided by the test context already bound
+            setattr(test, test._testMethodName, functools.partial(test_method, **matching_args))
```
In a past life, this would've been a bad code smell to me, but other test frameworks do similar things to bind arguments.

hidmic replied:
Yeah, I'd imagine `pytest` does something alike to bind its fixture objects as test functions' arguments.

hidmic reviewed (approved):
LGTM @pbaughman !

hidmic commented:
@wjwwood mind to take a look too? 

Also, for some reason I don't have full reviewer powers here.

pbaughman commented:
@hidmic Fixed your lack of review powers.  We missed you somehow when we added William
